File: sdk/src/types/intent.ts
import { z } from 'zod';

export enum IntentType {
  PAYIN = 'PAYIN',
  PAYOUT = 'PAYOUT',
  TRADE = 'TRADE',
}

export const StateHistoryEntrySchema = z.object({
  state: z.string(),
  timestamp: z.string(),
  reason: z.string().optional(),
});

export type StateHistoryEntry = z.infer<typeof StateHistoryEntrySchema>;

export const IntentSchema = z.object({
  id: z.string(),
  userId: z.string().optional(),
  intentType: z.string(),
  state: z.string(),
  amount: z.string(),
  currency: z.string(),
  actualAmount: z.string().optional(),
  referenceCode: z.string().optional(),
  bankTxId: z.string().optional(),
  chainId: z.string().optional(),
  txHash: z.string().optional(),
  stateHistory: z.array(StateHistoryEntrySchema).optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  expiresAt: z.string().optional(),
  completedAt: z.string().optional(),
  metadata: z.record(z.any()).optional(),
});

export type Intent = z.infer<typeof IntentSchema>;

export const VirtualAccountSchema = z.object({
  bank: z.string(),
  accountNumber: z.string(),
  accountName: z.string(),
});

export type VirtualAccount = z.infer<typeof VirtualAccountSchema>;

export const BankAccountSchema = z.object({
  bankCode: z.string(),
  accountNumber: z.string(),
  accountName: z.string(),
});

export type BankAccount = z.infer<typeof BankAccountSchema>;

export const CreatePayinRequestSchema = z.object({
  tenantId: z.string(),
  userId: z.string(),
  amountVnd: z.number(),
  railsProvider: z.string(),
  metadata: z.record(z.any()).optional(),
});

export type CreatePayinRequest = z.infer<typeof CreatePayinRequestSchema>;

export const CreatePayinResponseSchema = z.object({
  intentId: z.string(),
  referenceCode: z.string(),
  virtualAccount: VirtualAccountSchema.optional(),
  expiresAt: z.string(),
  status: z.string(),
});

export type CreatePayinResponse = z.infer<typeof CreatePayinResponseSchema>;

export const ConfirmPayinRequestSchema = z.object({
  tenantId: z.string(),
  referenceCode: z.string(),
  status: z.string(),
  bankTxId: z.string(),
  amountVnd: z.number(),
  settledAt: z.string(),
  rawPayloadHash: z.string(),
});

export type ConfirmPayinRequest = z.infer<typeof ConfirmPayinRequestSchema>;

export const ConfirmPayinResponseSchema = z.object({
  intentId: z.string(),
  status: z.string(),
});

export type ConfirmPayinResponse = z.infer<typeof ConfirmPayinResponseSchema>;

export const CreatePayoutRequestSchema = z.object({
  tenantId: z.string(),
  userId: z.string(),
  amountVnd: z.number(),
  railsProvider: z.string(),
  bankAccount: BankAccountSchema,
  metadata: z.record(z.any()).optional(),
});

export type CreatePayoutRequest = z.infer<typeof CreatePayoutRequestSchema>;

export const CreatePayoutResponseSchema = z.object({
  intentId: z.string(),
  status: z.string(),
});

export type CreatePayoutResponse = z.infer<typeof CreatePayoutResponseSchema>;

export const CreatePayInSchema = CreatePayinRequestSchema;
export type CreatePayInDto = CreatePayinRequest;
export const CreatePayOutSchema = CreatePayoutRequestSchema;
export type CreatePayOutDto = CreatePayoutRequest;

export const IntentFilterSchema = z.object({
  userId: z.string().optional(),
  intentType: z.string().optional(),
  state: z.string().optional(),
  limit: z.number().optional(),
  offset: z.number().optional(),
});

export type IntentFilters = z.infer<typeof IntentFilterSchema>;

File: sdk-python/src/rampos/models/intent.py
"""Intent models for pay-in and pay-out operations."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel


class IntentType(str, Enum):
    PAYIN = "PAYIN"
    PAYOUT = "PAYOUT"
    TRADE = "TRADE"


class StateHistoryEntry(BaseModel):
    state: str
    timestamp: str
    reason: str | None = None


class Intent(BaseModel):
    id: str
    user_id: str | None = None
    intent_type: str
    state: str
    amount: str
    currency: str
    actual_amount: str | None = None
    reference_code: str | None = None
    bank_tx_id: str | None = None
    chain_id: str | None = None
    tx_hash: str | None = None
    state_history: list[StateHistoryEntry] | None = None
    created_at: str
    updated_at: str
    expires_at: str | None = None
    completed_at: str | None = None
    metadata: dict[str, Any] | None = None

    model_config = {"populate_by_name": True, "alias_generator": None}


class VirtualAccount(BaseModel):
    bank: str
    account_number: str
    account_name: str


class BankAccount(BaseModel):
    bank_code: str
    account_number: str
    account_name: str


class CreatePayinRequest(BaseModel):
    tenant_id: str
    user_id: str
    amount_vnd: float
    rails_provider: str
    metadata: dict[str, Any] | None = None


class CreatePayinResponse(BaseModel):
    intent_id: str
    reference_code: str
    virtual_account: VirtualAccount | None = None
    expires_at: str
    status: str


class ConfirmPayinRequest(BaseModel):
    tenant_id: str
    reference_code: str
    status: str
    bank_tx_id: str
    amount_vnd: float
    settled_at: str
    raw_payload_hash: str


class ConfirmPayinResponse(BaseModel):
    intent_id: str
    status: str


class CreatePayoutRequest(BaseModel):
    tenant_id: str
    user_id: str
    amount_vnd: float
    rails_provider: str
    bank_account: BankAccount
    metadata: dict[str, Any] | None = None


class CreatePayoutResponse(BaseModel):
    intent_id: str
    status: str


class IntentFilters(BaseModel):
    user_id: str | None = None
    intent_type: str | None = None
    state: str | None = None
    limit: int | None = None
    offset: int | None = None
